### DISCUSSION TITLE
Confusing example and behavior regarding `If(...).Then(() => ...)`

### DISCUSSION THREAD

**Discussion by alexrp** [dotnet/dotNext#189] [Ideas]
The [docs](https://dotnet.github.io/dotNext/features/metaprogramming/lambda.html) have this example:

```csharp
Func<long, bool> isZero = Lambda<Func<long, bool>>(fun => 
{
    var arg = fun[0];
    If((Expression)(arg.AsDynamic() != 0L))
        .Then(() => Return(true))
        .Else(() => Return(false))
    .End();
}).Compile();
```

Is this doing what it's supposed to be doing? Trying this locally, `.Then(() => ...)` binds to `ConditionalBuilder.Then(Expression)`, *not* `CodeGenerator.Then(ConditionalBuilder, Action)`. This also makes the API a bit unwieldy in general, since I have to write code with casts like this:

```csharp
If(...)
    .Then((Action)(() =>
    {
        // ...
    }))
.End()
```

Not sure if I'm misunderstanding something or if this is unintended. 🤔

**Comment by sakno** [comment]
Thanks to MS, they introduced breaking change. See https://github.com/dotnet/csharplang/discussions/5157. If you have any suggestions, I'm ready to discuss.

**Reply by sakno** [reply]
Since C# introduces ambiguity, API can be changed in a verbose way: remove overloading for `Then` and `Else` and produce extra methods for explicit resolution, e.g.
```
Then().Block(Action);
Then().Expression(Expression);
```

**Reply by sakno** [reply]
Another interesting observation: C# differently resolves ambiguity between extension method and class method. I can replace `Then(Action)` extension method with class method declared directly in `ConditionalBuilder` class. In that case, C# resolves overload correctly.

**Reply by alexrp** [reply]
Interesting, I wasn't aware this was a breaking change in the language.

What do you think about `ThenBlock(Action)`/`ElseBlock(Action)` methods as a workaround? Similar idea to what you suggested, but a bit less verbose.

**Reply by sakno** [reply]
I've fixed this through moving extension methods to class methods. See e6c7499a40f182b7a8a275c408e1135d0cef0600. There is no good solution (without breaking changes). The change I made provides backward compatibility at source level (you need to rebuild the project w/o any changes) but not binary compatibility. `ThenBlock` is an option, but this is both source and binary incompatible change.

**Reply by sakno** [reply]
A new release 4.15.0 has been published. @alexrp , could you please mark this discussion as answered?

**Reply by alexrp** [reply]
I actually tried to do that earlier, but I don't see the option to do it for some reason. 🤔
